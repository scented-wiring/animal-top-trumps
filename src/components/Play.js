import "../styles/Play.css";
import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import axios from "axios";
import Alert from "../components/Alert";
import Card from "../components/Card";

const Play = () => {
  const initialState = {
    alert: {
      message: "",
      alertType: "",
    },
  };

  const [playerCards, setPlayerCards] = useState([]);
  const [playerCard, setPlayerCard] = useState({});
  const [computerCards, setComputerCards] = useState([]);
  const [computerCard, setComputerCard] = useState({});
  const [playerTurn, setPlayerTurn] = useState(true);
  const [alert, setAlert] = useState(initialState.alert);
  const [load, setLoad] = useState([true]);

  const shuffle = (array) => {
    var currentIndex = array.length,
      temporaryValue,
      randomIndex;

    while (0 !== currentIndex) {
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex -= 1;

      temporaryValue = array[currentIndex];
      array[currentIndex] = array[randomIndex];
      array[randomIndex] = temporaryValue;
    }

    return array;
  };

  const handleAlertPress = () => {
    setAlert({
      message: "",
      alertType: "",
    });
  };

  const playField = (field) => {
    if (playerCard[field] > computerCard[field]) {
      setPlayerCards((playerCards) => [...playerCards, computerCard]);
      computerCards.splice(0, 1);
      setComputerCards(computerCards);
      setComputerCard(computerCards[0]);
    } else {
      setComputerCards((computerCards) => [...computerCards, computerCard]);
      playerCards.splice(0, 1);
      setPlayerCards(playerCards);
      setPlayerCard(playerCards[0]);
    }
  };

  useEffect(() => {
    const deal = (array) => {
      let i;
      for (i = 0; i < array.length; i++) {
        if (array.indexOf(array[i]) % 2 === 0) {
          playerCards.push(array[i]);
        } else {
          computerCards.push(array[i]);
        }
      }
    };

    axios
      .get("http://localhost:3000/cards")
      .then((response) => {
        deal(shuffle(response.data));
        setPlayerCard(playerCards[0]);
        setComputerCard(computerCards[0]);
        setLoad(false);
      })
      .catch(() => {
        setAlert({
          message: "Could not connect to the server.",
          alertType: "Error",
        });
        setLoad(false);
      });
  }, []);

  if (!load) {
    return (
      <div className="Play">
        <div className="game-area">
          <div className="player">
            <div className="score">Player: {playerCards.length} cards</div>
            <Card {...playerCard} />
          </div>

          <div className="game-ui">
            {playerTurn ? (
              <div className="player-turn">
                <div className="turn">Player's Turn</div>
                <div className="select-field">
                  Select a field to play <br /> against your opponent:
                </div>
                <div className="field" onClick={() => playField("cool")}>
                  Cool
                </div>
                <div className="field" onClick={() => playField("largeness")}>
                  Largeness
                </div>
                <div className="field" onClick={() => playField("handsome")}>
                  Handsome
                </div>
              </div>
            ) : (
              <div className="turn">Computer's Turn</div>
            )}
          </div>

          <div className="computer">
            <div className="score">Computer: {computerCards.length} cards</div>
            <Card {...computerCard} secret={true} />
          </div>
        </div>
        <Alert
          message={alert.message}
          alertType={alert.alertType}
          onAlertPress={handleAlertPress}
        />
        <Link to="/">
          <h3>&larr; BACK</h3>
        </Link>
      </div>
    );
  } else {
    return (
      <div className="Play">
        <h2>Loading...</h2>
      </div>
    );
  }
};

export default Play;
