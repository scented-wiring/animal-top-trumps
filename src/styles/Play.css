.Play {
  display: flex;
  flex-direction: column;
  margin-top: 40px;
  align-items: center;
  width: 100%;
}

.game-area {
  display: flex;
  flex-direction: row;
  margin-bottom: 20px;
  width: 100%;
  justify-content: space-between;
}

.player {
  margin-left: 100px;
}

.computer {
  margin-right: 100px;
}

.score {
  display: flex;
  flex-direction: column;
  font-family: "Staatliches", sans-serif;
  font-size: 30px;
  text-align: center;
  margin-bottom: 10px;
}

.game-ui {
  justify-self: center;
  text-align: center;
}

.turn {
  font-family: "Staatliches", sans-serif;
  text-decoration: underline;
  font-size: 50px;
  margin-bottom: 20px;
}

.select-field {
  font-size: 25px;
  margin-bottom: 20px;
}

.field {
  font-family: "Staatliches", sans-serif;
  font-size: 50px;
  margin-bottom: 20px;
  transition: all 0.4s ease 0s;
}

.field:hover {
  color: red;
  cursor: pointer;
  transition: all 0.4s ease 0s;
  letter-spacing: 3px;
}
