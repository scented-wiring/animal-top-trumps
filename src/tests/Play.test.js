import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import Play from "../components/Play";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { getByText, getByRole, fireEvent } from "@testing-library/react";

const renderContainer = () => {
  render(
    <BrowserRouter>
      <Play />
    </BrowserRouter>,
    container
  );
};

const fakeResponse = [
  {
    id: 1,
    name: "Dog",
    cool: 1,
    largeness: 10,
    handsome: 9,
    aka: "Hairy friend",
    alignment: "Lawful Good",
  },
  {
    id: 2,
    name: "Cat",
    cool: 10,
    largeness: 10,
    handsome: 8,
    aka: "Furry assassin",
    alignment: "Lawful Evil",
  },
  {
    id: 3,
    name: "Elephant",
    cool: 8,
    largeness: 10,
    handsome: 6,
    aka: "Leg-face",
    alignment: "Lawful Neutral",
  },
  {
    id: 4,
    name: "Snake",
    cool: 2,
    largeness: 10,
    handsome: 4,
    aka: "Long lad",
    alignment: "Neutral Evil",
  },
];

const fakeResponse2 = [
  {
    id: 1,
    name: "Dog",
    cool: 1,
    largeness: 8,
    handsome: 9,
    aka: "Hairy friend",
    alignment: "Lawful Good",
  },
  {
    id: 2,
    name: "Cat",
    cool: 8,
    largeness: 10,
    handsome: 8,
    aka: "Furry assassin",
    alignment: "Lawful Evil",
  },
];

let container;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("Play component", () => {
  test("renders correctly", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(404);
    await act(async () => {
      renderContainer();
    });
    expect(container).toMatchSnapshot();
  });

  test("renders an alert if cannot connect to server", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(404);

    await act(async () => {
      renderContainer();
    });
    expect(getByText(container, "Error")).toBeInTheDocument();
    expect(
      getByText(container, "Could not connect to the server.")
    ).toBeInTheDocument();
  });

  test("deals out cards equally between player and computer", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);

    await act(async () => {
      renderContainer();
    });
    expect(getByText(container, "Player: 2 cards")).toBeInTheDocument();
    expect(getByText(container, "Computer: 2 cards")).toBeInTheDocument();
  });

  test("compares fields, determines winner and adjusts card arrays if one value exceeds another", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);

    await act(async () => {
      renderContainer();
    });

    fireEvent.click(getByText(container, "Cool"));
    expect(container.textContent).toContain("wins!");
    expect(container.textContent).toContain("3 cards");
    expect(container.textContent).toContain("1 cards");
  });

  test("compares fields, declares a tie and adjusts card arrays if values are equal", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);

    await act(async () => {
      renderContainer();
    });

    fireEvent.click(getByText(container, "Largeness"));
    expect(container.textContent).toContain("Tie!");
    expect(container.textContent).toContain("1 cards");
  });

  test("declares a winner if one card array is empty", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse2);

    await act(async () => {
      renderContainer();
    });

    fireEvent.click(getByText(container, "Cool"));
    fireEvent.click(getByText(container, "OK"));
    expect(container.textContent).toContain("wins the game!");
  });

  test("clears win message after confirmation click", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);

    await act(async () => {
      renderContainer();
    });

    fireEvent.click(getByText(container, "Cool"));
    expect(container.textContent).toContain("wins!");
    fireEvent.click(getByText(container, "OK"));
    expect(container.textContent).not.toContain("wins!");
  });

  test("changes score div class based on active turn", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);

    await act(async () => {
      renderContainer();
    });

    expect(getByText(container, "Player: 2 cards")).toHaveClass("score-active");
    expect(getByText(container, "Computer: 2 cards")).toHaveClass("score");
  });

  test("renders home link", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);

    await act(async () => {
      renderContainer();
    });

    const link = getByRole(container, "link");
    expect(link.textContent).toContain("← BACK");
    expect(link.getAttribute("href")).toBe("/");
  });
});
