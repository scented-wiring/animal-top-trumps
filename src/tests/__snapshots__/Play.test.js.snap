// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowseCards component renders correctly 1`] = `
<div>
  <div
    class="Play"
  >
    <div
      class="game-area"
    >
      <div
        class="player"
      >
        <div
          class="score"
        >
          Player: 
          2
           cards
        </div>
        <div
          class="Card-border"
        >
          <div
            class="Card"
          >
            <div
              class="name"
            >
              Elephant
            </div>
            <div
              class="other-stat"
            >
              Cool: 
               
            </div>
            <div
              class="other-stat"
            >
              Largeness: 
               
            </div>
            <div
              class="other-stat"
            >
              Handsome: 
               
            </div>
            <div
              class="other-stat"
            >
              aka: 
              <br />
              "
              "
               
            </div>
            <div
              class="other-stat"
            >
              Alignment:
              <br />
               
            </div>
          </div>
        </div>
      </div>
      <div
        class="game-ui"
      >
        <div
          class="player-turn"
        >
          <div
            class="turn"
          >
            Player's Turn
          </div>
          <div
            class="select-field"
          >
            Select a field to play 
            <br />
             against your opponent:
          </div>
          <div
            class="field"
          >
            Cool
          </div>
          <div
            class="field"
          >
            Largeness
          </div>
          <div
            class="field"
          >
            Handsome
          </div>
        </div>
      </div>
      <div
        class="computer"
      >
        <div
          class="score"
        >
          Computer: 
          1
           cards
        </div>
        <div
          class="hidden-card"
        >
          <div
            class="text"
          >
            Animal 
            <br />
            Top Trumps
          </div>
          <img
            alt="Logo"
            height="96px"
            id="card-logo"
            src="dog.png"
            width="96px"
          />
        </div>
      </div>
    </div>
    <a
      href="/"
    >
      <h3>
        ‚Üê BACK
      </h3>
    </a>
  </div>
</div>
`;
